src/bestcrypt_ve_fmt_plug.c

```diff
@@ -0,0 +1,456 @@
+
+	MEM_FREE(saved_key);
+	MEM_FREE(cracked);
+	MEM_FREE(saved_len);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *p = ciphertext, *ctcopy, *keeptr;
+	int extra;
+
+	if (strncmp(ciphertext, FORMAT_TAG, FORMAT_TAG_LENGTH))
+		return 0;
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LENGTH;
+	if ((p = strtokm(ctcopy, "$")) == NULL) // version
+		goto err;
+	if (!isdec(p))
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL) // encryption alogrithm
+		goto err;
+	if (!isdec(p))
+		goto err;
+	if (atoi(p) != 8 && atoi(p) != 9 && atoi(p) != 10 && atoi(p) != 11 && atoi(p) != 15)
+		goto err;
+	if (atoi(p) == 11) {
+		fprintf(stderr, "RC6 encryption not supported yet!\n");
+		goto err;
+	}
+	if ((p = strtokm(NULL, "$")) == NULL) // salt
+		goto err;
+	if (hexlenl(p, &extra) != 48 || extra)
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL) // encrypted content
+		goto err;
+	if (hexlenl(p, &extra) != 0x60 * 2 || extra)
+		goto err;
+
+	MEM_FREE(keeptr);
+	return 1;
+
+err:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+static void *get_salt(char *ciphertext)
+{
+	static struct custom_salt cs;
+	int i;
+	char *p = ciphertext, *ctcopy, *keeptr;
+	memset(&cs, 0, sizeof(cs));
+
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LENGTH;
+	p = strtokm(ctcopy, "$");
+	cs.version = atoi(p);
+	p = strtokm(NULL, "$");
+	cs.enc_algoID = atoi(p);
+	p = strtokm(NULL, "$");
+	for (i = 0; i < 24; i++)
+		cs.salt[i] = (atoi16[ARCH_INDEX(p[2*i])] << 4) | atoi16[ARCH_INDEX(p[2*i+1])];
+	p = strtokm(NULL, "$");
+	for (i = 0; i < 0x60; i++)
+		cs.encrypted_data[i] = (atoi16[ARCH_INDEX(p[2*i])] << 4) | atoi16[ARCH_INDEX(p[2*i+1])];
+
+	MEM_FREE(keeptr);
+
+	return (void *)&cs;
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static void bcve_set_key(char *key, int index)
+{
+	strnzcpy(saved_key[index], key, sizeof(saved_key[index]));
+}
+
+static char *get_key(int index)
+{
+	return saved_key[index];
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index;
+	int failed = 0;
+
+	memset(cracked, 0, sizeof(cracked[0])*cracked_count);
+
+	yescrypt_params_t params = { .N = 0x8000, .r = 16, .p = 1 };
+
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index++) {
+		unsigned char kdf_out[32];
+		/* BestCrypt uses CBC mode with a null IV */
+		unsigned char iv[16] = {0};
+		unsigned char out[0x60];
+		unsigned char intermediate_keys[64];
+
+		SHA256_CTX ctx;
+		unsigned char sha256_hash[32];
+
+#ifdef _OPENMP
+		int t = omp_get_thread_num();
+		if (t >= max_threads) {
+			failed = -1;
+			continue;
+		}
+#else
+		const int t = 0;
+#endif
+		yescrypt_kdf(NULL, &local[t],
+		             (const uint8_t *)saved_key[index],
+		             strlen(saved_key[index]),
+		             (const uint8_t *)cur_salt->salt,
+		             0x18,
+		             &params,
+		             kdf_out, 0x20);
```

Comment:
Must not ignore the return value from `yescrypt_kdf`. Please see `restic_fmt_plug.c` for how to handle it.

Can also use `sizeof` for the 0x18 and 0x20 here.